### PR TITLE
featurizer

### PR DESCRIPTION
**Pull request by nzarayeneh**

### CODE REVIEW

**Comment by parkererickson** [comment]
Do we want to add a `timeout` parameter in `runAlgorithm()` that is passed into the `runInstalledQuery()` call? Might be helpful, we can set the default timeout as a large one if it is not passed in.

**Comment by nzarayeneh** [comment]
Yeah that is a good idea, I just added it. We can set the default to be
2147480 or any value which makes more sense, but 16 is too low I guess.

Thanks,

File: pyTigerGraph/pyTigerGraph.py [hunk 1]
```diff
@@ -22,12 +23,13 @@ class TigerGraphConnection(pyTigerGraphVertex, pyTigerGraphEdge, pyTigerGraphUDT
     pyTigerGraphLoading, pyTigerGraphPath):
     """Python wrapper for TigerGraph's REST++ and GSQL APIs"""
 
-    def __init__(self, host: str = "http://127.0.0.1", graphname: str = "MyGraph",
+    def __init__(self, host: str = "http://34.232.63.121", graphname: str = "BankSim",
```
Comment (tg-bill) [inline pyTigerGraph/pyTigerGraph.py:26]: We don't want to change this default. The new address is only relevant to us but not to a general user.

File: pyTigerGraph/gds/featurizer.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        query = ("USE GRAPH {}\n".format(self.conn.graphname) + query + "\ninstall Query {}\n".format(name_of_query))
+        print("Installing and optimizing the queries, it might take a minute")
+        resp = self.conn.gsql(query)
+        status = resp.splitlines()[-1]
+        if "Failed" in status:
+            raise ConnectionError(status)
+        return name_of_query 
+
+    def installAlgorithm(self,query_name:str,schema_type:str="VERTEX",attr_name:str=None,schema_name:str=None):
+        query_path = pjoin(os.path.dirname(os.path.abspath(__file__)), "gsql", "featurizer", query_name+'.gsql')
```
Comment (tg-bill) [inline pyTigerGraph/gds/featurizer.py:84]: Might want to check if the input query_name is in the allowlist of query names

**Comment by parkererickson** [comment]
With the dataloaders, Bill wrote it in such a way so you could call `conn.gds.neighborLoader()` and that would return the appropriate dataloader. Could we do something similar to avoid `f = conn.gds.featurizer(conn)` or whatever was the exact syntax that was demoed?

File: pyTigerGraph/gds/featurizer.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+import re
+import random
+import string
+
+class Featurizer:
+    '''Install and run different graph algorithms available in graph data science library.'''
+    def __init__(
+    self, 
+    conn: "TigerGraphConnection",
+    name_of_query: str = None,
```
Comment (tg-bill) [inline pyTigerGraph/gds/featurizer.py:24]: Looks like this input is not used?

File: pyTigerGraph/gds/featurizer.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            if 'v_type' in params.values():
+                params['v_type'] = self.conn.getVertexTypes()[0]
+            if 'e_type' in params.values():
+                params['e_type'] = self.conn.getEdgeTypes()[0]
+            print("Default parameters are:",params)
+        if None in params.values():
+            raise ValueError("Query parameters which are None need to be initialized.")
+        result = self.conn.runInstalledQuery(name_of_query, params,timeout=timeout)
+        if result != None:
+            return "Success!"
```
Comment (tg-bill) [inline pyTigerGraph/gds/featurizer.py:184]: As @parkererickson mentioned, might want to return the result.

File: pyTigerGraph/gds/gsql/featurizer/degrees.gsql [hunk 5]
```diff
@@ -0,0 +1,8 @@
+CREATE QUERY degrees()  SYNTAX V1{
```
Comment (tg-bill) [inline pyTigerGraph/gds/gsql/featurizer/degrees.gsql:1]: Looks like this function is not included in the allowlist. What is it for?

File: pyTigerGraph/gds/gsql/featurizer/tg_label_prop.gsql [hunk 6]
```diff
@@ -0,0 +1,66 @@
+CREATE QUERY tg_label_prop (SET<STRING> v_type, SET<STRING> e_type, INT max_iter=10, INT output_limit=10, 
+  BOOL print_accum = TRUE, STRING file_path = "", STRING attr = "")  SYNTAX V1 {
```
Comment (tg-bill) [inline pyTigerGraph/gds/gsql/featurizer/tg_label_prop.gsql:2]: Could you try if V2 works? If so let's use V2 as the database is pushing for V2

File: pyTigerGraph/gds/gsql/featurizer/tg_pagerank.gsql [hunk 7]
```diff
@@ -0,0 +1,74 @@
+CREATE QUERY tg_pagerank (STRING v_type, STRING e_type,
+ FLOAT max_change=0.001, INT max_iter=25, FLOAT damping=0.85, INT top_k = 100,
+ BOOL print_accum = TRUE, STRING result_attr =  "", STRING file_path = "",
+ BOOL display_edges = FALSE) SYNTAX V1 {
```
Comment (tg-bill) [inline pyTigerGraph/gds/gsql/featurizer/tg_pagerank.gsql:4]: Same as above.

**Comment by nzarayeneh** [comment]
Bill has used a BaseLoader class as he has several loader, if I want to do
something similar I have to define a Featurizer class and consider two
functions (installAlgorithm and runAlgorithm) as two classes, that is
possible if we think this way is better

**Review by billshitg — COMMENTED** [review]
Looks good overall. Thanks for the nice work. I left detailed feedbacks in the code. Could you take a look at them? I am being very nitpicking so feel free to say no to any of them.

**Comment by parkererickson** [comment]
> With the dataloaders, Bill wrote it in such a way so you could call conn.gds.neighborLoader() and that would return the appropriate dataloader. Could we do something similar to avoid f = conn.gds.featurizer(conn) or whatever was the exact syntax that was demoed?

Determined to work without passing in the connection.

**Comment by parkererickson** [comment]
Last thing from me would be to add detailed docstrings to the various function headers. See other files for format, it helps with generating the documentation automatically. Thanks!

File: pyTigerGraph/gds/gsql/featurizer/degrees.gsql (hunk 5 above)
Comment (nzarayeneh) [reply pyTigerGraph/gds/gsql/featurizer/degrees.gsql:1]: Just to check when the algorithm does not have any input parameters

File: pyTigerGraph/gds/gsql/featurizer/tg_label_prop.gsql (hunk 6 above)
Comment (nzarayeneh) [reply pyTigerGraph/gds/gsql/featurizer/tg_label_prop.gsql:2]: Yes, it is working I just changed it

File: pyTigerGraph/gds/gsql/featurizer/tg_pagerank.gsql (hunk 7 above)
Comment (nzarayeneh) [reply pyTigerGraph/gds/gsql/featurizer/tg_pagerank.gsql:4]: Done

File: pyTigerGraph/pyTigerGraph.py (hunk 1 above)
Comment (nzarayeneh) [reply pyTigerGraph/pyTigerGraph.py:26]: I changed it

File: pyTigerGraph/gds/featurizer.py (hunk 2 above)
Comment (nzarayeneh) [reply pyTigerGraph/gds/featurizer.py:84]: added the check

File: pyTigerGraph/gds/featurizer.py (hunk 4 above)
Comment (nzarayeneh) [reply pyTigerGraph/gds/featurizer.py:184]: I changed accordingly

File: pyTigerGraph/gds/gsql/featurizer/degrees.gsql (hunk 5 above)
Comment (tg-bill) [reply pyTigerGraph/gds/gsql/featurizer/degrees.gsql]: got it. but I see there is such a file in the tests folder. Is this one a duplicate?

**Review by billshitg — APPROVED** [review]

File: pyTigerGraph/gds/gsql/featurizer/degrees.gsql (hunk 5 above)
Comment (nzarayeneh) [reply pyTigerGraph/gds/gsql/featurizer/degrees.gsql]: yeah, that should be a duplicate. I created the same gsql folder in test as well, but we do not need that

Comment (tg-bill) [reply pyTigerGraph/gds/gsql/featurizer/degrees.gsql]: cool. thanks.

**Pull request merged**